clippy_lints/src/mut_mutex_lock.rs

```diff
@@ -0,0 +1,75 @@
+use crate::utils::{is_type_diagnostic_item, span_lint_and_help};
+use if_chain::if_chain;
+use rustc_hir::{Expr, ExprKind, Mutability};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for `&mut Mutex::lock` calls
+    ///
+    /// **Why is this bad?** `Mutex::lock` is less efficient than
+    /// calling `Mutex::get_mut`
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// use std::sync::{Arc, Mutex};
+    ///
+    /// let mut value_rc = Arc::new(Mutex::new(42_u8));
+    /// let value_mutex = Arc::get_mut(&mut value_rc).unwrap();
+    ///
+    /// let mut value = value_mutex.lock().unwrap();
+    /// *value += 1;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::sync::{Arc, Mutex};
+    ///
+    /// let mut value_rc = Arc::new(Mutex::new(42_u8));
+    /// let value_mutex = Arc::get_mut(&mut value_rc).unwrap();
+    ///
+    /// let value = value_mutex.get_mut().unwrap();
+    /// *value += 1;
+    /// ```
+    pub MUT_MUTEX_LOCK,
+    correctness,
+    "`&mut Mutex::lock` does unnecessary locking"
+}
+
+declare_lint_pass!(MutMutexLock => [MUT_MUTEX_LOCK]);
+
+impl<'tcx> LateLintPass<'tcx> for MutMutexLock {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, ex: &'tcx Expr<'tcx>) {
+        if_chain! {
+            if is_mut_mutex_lock_call(cx, ex).is_some();
+            then {
+                span_lint_and_help(
+                    cx,
+                    MUT_MUTEX_LOCK,
+                    ex.span,
+                    "calling `&mut Mutex::lock` unnecessarily locks an exclusive (mutable) reference",
+                    None,
+                    "use `&mut Mutex::get_mut` instead",
+                );
+            }
+        }
+    }
+}
+
+fn is_mut_mutex_lock_call<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<&'tcx Expr<'tcx>> {
```

Comment:
[IMO] I think it would be better to return `bool` in the method that has `is_` prefix. Also, that looks good enough here.
```suggestion
fn is_mut_mutex_lock_call<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> bool {
```

Reply:
Alright. If I do change the signature from `Option` to `bool`, I get a `needless_bool` error demonstrated in the snippet below: 

```rust
if_chain! {
  then { true } 
  else { false }
}
```

I could do the following:

```rust
let opt = if_chain! { 
  then { Some(_) }
  else { None }
};

opt.is_some()
```

It seems a bit awkward. I could just inline the function instead:

```rust
fn check_expr() {
  if_chain! {
    then {
     span_lint_and_help();
    }
  }
}
```

Is that okay?

Reply:
Thanks for your comments!
I think it looks good to inline the function.

[FYI] We could write like below.
```
if_chain! {
        ... (snip)
        if let ty::Ref(_, inner_ty, Mutability::Mut) = ty.kind();
        then {
            is_type_diagnostic_item(cx, inner_ty, sym!(mutex_type))
        } else {
            false
        }
    }
```

or https://github.com/rust-lang/rust-clippy/blob/28f9b84042b7d0ffd1f913e8ba82348ebe23dd1b/clippy_lints/src/identity_op.rs#L61-L73

Reply:
Oh right, you could just return the last condition in the `if_chain` instead. 

Anyway, inlined the function instead.